testsuite/classlibrary/TestUGen_RTAlloc.sc

```diff
@@ -1,326 +1,325 @@
-				);
-			} {
-				this.assert(false, "% allocFail test should complete".format(name.asString));
-			}
-		};
-
-		// special test for PartConv: requires buffer alloc
-		var partConvTest = {
-			var b = Buffer.loadCollection(server, 1!(2**9));
-			var testFunc = { PartConv.ar(DC.ar(1), 2**8, b) };
-			var memFill;
-			server.sync;
-			testAllocPass.value("PartConv", testFunc, 0.1);
-			memFill = fillMemFunc.play(server);
-			testAllocFail.value("PartConv", testFunc, 0.1);
-			memFill.free;
+	test_allocFail_clearUnit {
+		var output = this.awaitSynthOutput {
+			DelayN.ar(DC.ar(1), this.memSizeSeconds * 2)
 		};
+		this.assertEquals(output[0], 0.0,
+			"a failed RTAlloc should clear unit outputs")
+	}
 
-		var memFill;
-
-		// RUN TESTS
-
-		ugenTests.keysValuesDo {|name, func|
-			testAllocPass.value(name.asString, func, 0.001);
+	test_allocFail_continueProcessing_sameNode {
+		var output = this.awaitSynthOutput {
+			DelayN.ar(DC.ar(1), this.memSizeSeconds * 2); DC.ar(1)
 		};
+		this.assertEquals(output[0], 1.0,
+			"a failed RTAlloc should not block other units in the same node")
+	}
 
-		onlyPassUGens.keysValuesDo {|name, func|
-			testAllocPass.value(name.asString, func, 0.001);
-		};
+	test_allocFail_continueProcessing_otherNodes {
+		var output = nil;
+		var dummyOutBus = Bus.audio(server, 1);
+		var memFill = { DelayN.ar(Silent.ar, this.memSizeSeconds * 2) }.play(server, dummyOutBus);
+		server.sync;
+		output = this.awaitSynthOutput { DC.ar(1) };
+		memFill.free; dummyOutBus.free;
+		this.assertEquals(output[0], 1.0,
+			"a failed RTAlloc should not block other nodes in the processing chain")
+	}
 
-		longerRuntimeUGens.keysValuesDo {|name, func|
-			testAllocPass.value(name.asString, func, 0.02);
+	test_allocFail_setDoneFlag {
+		var output = this.awaitSynthOutput {
+			K2A.ar(Done.kr(DelayN.ar(Silent.ar, this.memSizeSeconds * 2)))
 		};
+		this.assert(output[0] > 0,
+			"a failed RTAlloc should set the Done flag")
+	}
 
-		// now fill mem and test allocFail
-		memFill = fillMemFunc.play(server);
+	// TEST ALL UGENS
+	// for each UGen in /server/plugins which calls RTAlloc
+	// - test successful alloc: should output non-zero values
+	// - test failed alloc: should only output zeros
 
-		ugenTests.keysValuesDo {|name, func|
-			testAllocFail.value(name.asString, func, 0.001);
-		};
+	// NOTE: we try to keep runtime for each test to the bare minimum, which varies
+	// from UGen to UGen (some need only 1 sample, some need 4 or more)
+	// Runtimes for each UGen are kept the same in both allocPass and allocFails tests
 
-		longerRuntimeUGens.keysValuesDo {|name, func|
-			testAllocFail.value(name.asString, func, 0.02);
-		};
+	// FFT UGens needs separate testing (see below)
+	// Some UGens were left out (see end of file)
 
-		memFill.free;
-		// now test PartConv
-		partConvTest.value;
+	test_allUGens_allocPass {
+		// allocate as little memory as possible for each UGen
+		var allocSeconds = 1 / server.sampleRate; // 1 sample
```

Comment:
This seems like a rather tight duration default. Are there any potential downsides, esp. wrt FFT ugens which are using some minimum number of samples for their FFT buffer?

In general, if the tests are running in parallel, having a minimum test dur wth a bit more overhead (even, say, 1 block) shouldn't be a big hit for overall test time?